[quant][graphmode] Add extra filtering for scale/zero_point/dtype in FoldQuantizeCallIntoBuffer

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26224 [quant][graphmode] Add extra filtering for scale/zero_point/dtype in FoldQuantizeCallIntoBuffer**
* #26223 [jit] Add filter function to subgraph rewriter runGraph

Summary:
We need to make sure they are Constant before we can do folding

Test Plan:
python test/test_jit.py 'TestJit.test_fold_quantize'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17462530](https://our.internmc.facebook.com/intern/diff/D17462530)

jerryzh168 committed:
[quant][graphmode] Add extra filtering for scale/zero_point/dtype in FoldQuantizeCallIntoBuffer

Summary:
We need to make sure they are Constant before we can do folding

Test Plan:
python test/test_jit.py 'TestJit.test_fold_quantize'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Add extra filtering for scale/zero_point/dtype in FoldQuantizeCallIntoBuffer"


Summary:
We need to make sure they are Constant before we can do folding

Test Plan:
python test/test_jit.py 'TestJit.test_fold_quantize'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
Update on "[quant][graphmode] Add extra filtering for scale/zero_point/dtype in FoldQuantizeCallIntoBuffer"


Summary:
We need to make sure they are Constant before we can do folding

Test Plan:
python test/test_jit.py 'TestJit.test_fold_quantize'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add extra filtering for scale/zero_point/dtype in FoldQuantizeCallIntoBuffer"


Summary:
We need to make sure they are Constant before we can do folding

Test Plan:
python test/test_jit.py 'TestJit.test_fold_quantize'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -937,7 +937,19 @@ graph(%self, %scale, %zero_point, %dtype):
   auto method = module.get_method(method_name);
   auto graph = method.graph();
   auto matches = findPatternMatches(pattern_graph, *graph);
+  auto filter = [](const Match& match, const std::unordered_map<std::string, Value*>& vmap) {
```
I think it's worth adding a comment here.

ZolotukhinM reviewed (approved):
Looks good!

jerryzh168 commented on torch/csrc/jit/passes/quantization.cpp:
```diff
@@ -937,7 +937,19 @@ graph(%self, %scale, %zero_point, %dtype):
+        dtype_node->kind() == prim::Constant;
+  };
   for (const auto& match : matches) {
+    if (!filter(match, vmap)) {
```
we don't have the same filter to skip the deleted nodes(https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/passes/subgraph_rewrite.cpp#L62), is that a concern?

jerryzh168 committed:
Update on "[quant][graphmode] Add extra filtering for scale/zero_point/dtype in FoldQuantizeCallIntoBuffer"


Summary:
We need to make sure they are Constant before we can do folding

Test Plan:
python test/test_jit.py 'TestJit.test_fold_quantize'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add extra filtering for scale/zero_point/dtype in FoldQuantizeCallIntoBuffer"


Summary:
We need to make sure they are Constant before we can do folding

Test Plan:
python test/test_jit.py 'TestJit.test_fold_quantize'

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D17462530](https://our.internmc.facebook.com/intern/diff/D17462530)

[ghstack-poisoned]
